lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -512,8 +512,84 @@ class ElementwiseOpConversionBase
   using OpAdaptor = typename SourceOp::Adaptor;
 
   explicit ElementwiseOpConversionBase(
-      TritonGPUToLLVMTypeConverter &typeConverter, PatternBenefit benefit = 1)
-      : ConvertTritonGPUOpToLLVMPattern<SourceOp>(typeConverter, benefit) {}
+      TritonGPUToLLVMTypeConverter &typeConverter,
+      ModuleAxisInfoAnalysis &axisAnalysisPass,
+      PatternBenefit benefit = 1)
+      : ConvertTritonGPUOpToLLVMPattern<SourceOp>(typeConverter, benefit),
+        axisAnalysisPass(axisAnalysisPass) {}
+
+  // Try to deduplicate the resultVals based on the
+  // constancy properties of the result discovered by
+  // the axis analysis pass. If possible, redundant
+  // computation is eliminated.
+  SmallVector<Value> maybeDeduplicate(SourceOp op,
+                                      SmallVector<Value> resultVals) const {
+    if (!isMemoryEffectFree(op))
+      // the op has side effects: can't dedup
+      return resultVals;
+    SmallVector<Value> results = op->getResults();
+    if (results.size() == 0 || results.size() > 1)
+      // there must be exactly 1 result
+      return resultVals;
+    Value result = results[0];
+    Type type = result.getType();
+    if (!type)
+      return resultVals;
+    RankedTensorType rtType = type.dyn_cast<RankedTensorType>();
+    if (!rtType)
+      // the result must be a tensor
+      return resultVals;
+
+    SmallVector<unsigned> elemsPerThread =
+        triton::gpu::getElemsPerThread(rtType);
+    int rank = elemsPerThread.size();
```

Comment:
This isn't the actual `rank` of a tensor

Reply:
Or do you mean the heuristics do not apply when the number of elements is greater than the number of rank? Why?

Reply:
Would invoking `.size()` here not get the rank? My understanding was that `elemsPerThread.size()` and `constancy.size()` (as well as `rtType.getShape().size()` not used here) should all return the same number: the rank (the number of dimensions) of the tensor. Is this not true?

Reply:
Assuming that the rank (i.e., the `.size()`) of the `elemsPerThread` and `constancy` is the same, the downstream logic is to check that, in each of the dimensions in the rank, one of those is divisible by the other:

1. If the `elemsPerThread[i]` is divisible by the `constancy[i]`, we can partially deduplicate the values by replicating a single value in each `constancy[i]`-sized group `constancy[i] - 1` times.

2. If the `constancy[i]` is divisible by the `elemsPerThread[i]`, this means that, along the `i`th dimension, the constancy covers the whole bunch of elements in the thread and we can deduplicate them all by reusing the first value `elemsPerThread[i] - 1` times.

If neither of those conditions hold, we can't deduplicate, as there is no even coverage of `elemsPerThreads` by the `constancy` in one or more dimensions. Does the above make sense?

Reply:
> Would invoking .size() here not get the rank? My understanding was that elemsPerThread.size() and constancy.size() (as well as rtType.getShape().size() not used here) should all return the same number: the rank (the number of dimensions) of the tensor. Is this not true?

It's not true. `elemsPerThreads` could be more than the number of rank

Reply:
Oh, I guess I misunderstood the code. 

`elemsPerThreads.size()` should be the same as the rank. I read is as `elemsPerThreads`